include/sdf/Root.hh

```diff
@@ -151,6 +158,29 @@ namespace sdf
     /// not been called.
     public: sdf::ElementPtr Element() const;
 
+    /// \brief Add a world to the root.
+    /// \param[in] _word World to add.
+    /// \return True if successful, false if a world with the name already
+    /// exists.
+    /// \return Errors, which is a vector of Error objects. Each Error includes
+    /// an error code and message. An empty vector indicates no error.
+    public: Errors AddWorld(const World &_world);
+
+    /// \brief Remove all worlds.
+    public: void ClearWorlds();
+
+    /// \brief Deep copy the provided Root object into this Root object.
+    /// \param[in] _root The Root object to copy.
+    /// Deprecate this function in SDF version 13, and use
+    /// IGN_UTILS_IMPL_PTR instead.
+    public: void Clone(const sdf::Root &_root);
```

Comment:
Wow. Learned something new. My reading of the `IGN_UTILS_UNIQUE_IMPL_PTR` indicated that copying was not allowed. But, the `Root` class did not mark the copy constructor as deleted....so we can copy. Feels like we may have abused the IGN_UTILS_UNIQUE_IMPL_PTR` in this case, unless this pattern was done on purpose.

Reply:
Change in [ab3d64d](https://github.com/ignitionrobotics/sdformat/pull/841/commits/ab3d64d0ec05357f65dcba19150ba8869af43517)

Reply:
If I add this:
```
public: Root(const Root &_root) = delete;
```

Then the code won't compile, with the following:
```
home/nkoenig/work/ignition/fortress/sdformat-root-mutable/src/Root.cc:455:10: error: use of deleted function ‘sdf::v12::Root::Root(const sdf::v12::Root&)’
  455 |   return r;
n file included from /home/nkoenig/work/ignition/fortress/sdformat-root-mutable/src/Root.cc:26:
/home/nkoenig/work/ignition/fortress/sdformat-root-mutable/include/sdf/Root.hh:58:13: note: declared here
   58 |     public: Root(const Root &_root) = delete;
      |             ^~~~
```

The documentation in ign-utils indicates that users of `IGN_UTILS_UNIQUE_IMPLE_PTR` should delete the copy constructor. See here: https://github.com/ignitionrobotics/ign-utils/blob/ign-utils1/include/ignition/utils/ImplPtr.hh#L192

I would assume the default copy constructor is being used here.

Reply:
hmm, that's very interesting. Without making any changes to the PR, if you do this:
```
sdf::Root root1;
sdf::Root root2(root1);
```
the code won't compile:

```
/home/addisu/ws/fortress/src/sdformat/src/Root_TEST.cc:326:24: error: use of deleted function ‘sdf::v12::Root::Root(const sdf::v12::Root&)’
  326 |   sdf::Root root2(root1);
      |                        ^
In file included from /home/addisu/ws/fortress/src/sdformat/src/Root_TEST.cc:28:
/home/addisu/ws/fortress/src/sdformat/include/sdf/Root.hh:54:26: note: ‘sdf::v12::Root::Root(const sdf::v12::Root&)’ is implicitly deleted because the default definition would be ill-formed:
   54 |   class SDFORMAT_VISIBLE Root
      |                          ^~~~
/home/addisu/ws/fortress/src/sdformat/include/sdf/Root.hh:54:26: error: use of deleted function ‘std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = sdf::v12::Root::Implementation; _Dp = void (*)(sdf::v12::Root::Implementation*)]’
```

so I guess, explicitly deleting the copy ctor disables RVO?

Reply:
Investigating this further, explicitly deleting the copy ctor implicitly deletes the move ctor. Defaulting the move ctor fixes the error.

```
    public: Root(const Root &) = delete;
    public: sdf::Root & operator=(const Root &) = delete;

    public: Root(Root &&) = default;
    public: sdf::Root & operator=(Root &&) = default;
```

From what I can understand, we are using named return value optimization (NRVO), which is not [guaranteed](https://en.cppreference.com/w/cpp/language/copy_elision#Non-mandatory_elision_of_copy.2Fmove_.28since_C.2B.2B11.29_operations) in C++17 (unlike unnamed RVO), so the compiler might use either the [move or copy ctors ](https://en.cppreference.com/w/cpp/language/return#Notes)when returning from `Clone`. If we delete both (even implicitly), then the build fails.